kobo/apps/audit_log/tests/test_models.py

```diff
@@ -192,3 +202,120 @@ def test_create_auth_log_with_extra_metadata(
+    def test_create_submission_auth_log(
+        self, patched_create, patched_ip, patched_source
+    ):
+        request = self._create_request(
+            reverse('submissions'),
+            AnonymousUser(),
+            AccessLogModelTestCase.super_user,
+        )
+        with skip_all_signals():
+            AccessLog.create_from_request(request)
+        # just test that we delegated the object creation to the SubmissionAccessLog model, which
+        # itself is tested later
+        patched_create.assert_called_once_with(
+            user=AccessLogModelTestCase.super_user,
+            metadata={
+                'ip_address': '127.0.0.1',
+                'source': 'source',
+                'auth_type': ACCESS_LOG_SUBMISSION_AUTH_TYPE,
+            },
+        )
+
+
+class SubmissionAccessLogModelTestCase(BaseAuditLogTestCase):
+
+    fixtures = ['test_data']
+
+    def test_create_submission_access_log(self):
+        user = User.objects.get(username='someuser')
+        with skip_all_signals():
+            log = SubmissionAccessLog.objects.create(
+                user=user, metadata={'foo': 'bar'}
+            )
+        self.assertIsInstance(log, SubmissionAccessLog)
+        self._check_common_fields(log, user)
+        self.assertDictEqual(
+            log.metadata,
+            {'auth_type': ACCESS_LOG_SUBMISSION_AUTH_TYPE, 'foo': 'bar'},
+        )
+
+    def test_create_submission_access_log_overrides_auth_type(self):
+        user = User.objects.get(username='someuser')
+        with skip_all_signals():
+            log = SubmissionAccessLog.objects.create(
+                user=user, metadata={'auth_type': 'Token'}
+            )
+        self.assertEqual(
+            log.metadata['auth_type'], ACCESS_LOG_SUBMISSION_AUTH_TYPE
+        )
+
+    def test_add_self_to_existing_group(self):
+        user = User.objects.get(username='someuser')
+        with skip_all_signals():
+            group: SubmissionGroup = SubmissionGroup.objects.create(user=user)
+            log: SubmissionAccessLog = SubmissionAccessLog.objects.create(
+                user=user
+            )
+            log.add_to_existing_submission_group(group)
+            log.save()
+        self.assertEqual(log.submission_group.id, group.id)
+
+    @patch('kobo.apps.audit_log.models.logging.error')
+    def test_cannot_add_self_to_existing_group_with_wrong_user(
+        self, patched_error
+    ):
+        user1 = User.objects.get(username='someuser')
+        user2 = User.objects.get(username='anotheruser')
+        with skip_all_signals():
+            group: SubmissionGroup = SubmissionGroup.objects.create(user=user1)
+            log: SubmissionAccessLog = SubmissionAccessLog.objects.create(
+                user=user2
+            )
+            log.add_to_existing_submission_group(group)
+            log.save()
+        # make sure we reported an error
+        patched_error.assert_called_once()
```

Comment:
This should never happen in a real scenario. It's for future-proofing in case other code needs to use this method. I wasn't sure not adding to a group would be a big enough problem to raise an exception but I'm pretty ambivalent about it